pkg/scheduler/core/extender_test.go

```diff
@@ -112,20 +113,80 @@ type FakeExtender struct {
 	filteredNodes    []*v1.Node
 }
 
+// selectVictimsOnNodeByExtender checks the given nodes->pods map with predicates on extender's side.
+// Returns:
+// 1. More victim pods (if any) amended by preemption phase of extender.
+// 2. Number of violating victim (used to calculate PDB).
+// 3. Fits or not after preemption phase on extender's side.
+func (f *FakeExtender) selectVictimsOnNodeByExtender(
+	pod *v1.Pod,
+	node *v1.Node,
+	originVictims *schedulerapi.Victims,
+	pdbs []*policy.PodDisruptionBudget,
+) ([]*v1.Pod, int, bool) {
+	if fits, _ := f.runPredicate(pod, node); !fits {
+		return nil, 0, false
+	}
+
+	return []*v1.Pod{}, 0, true
+}
+
+func (f *FakeExtender) SupportsPreemption() bool {
+	return true
+}
+
+func (f *FakeExtender) ProcessPreemption(
+	pod *v1.Pod,
+	nodeToVictims map[string]*schedulerapi.Victims,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+) (schedulerapi.ExtenderPreemptionResult, error) {
+	nodeToVictimsCopy := map[string]*schedulerapi.Victims{}
+	// We don't want to change the original nodeToVictims
+	for k, v := range nodeToVictims {
+		nodeToVictimsCopy[k] = v
+	}
+
+	for nodeName, victims := range nodeToVictimsCopy {
```

Comment:
The reason we abandoned this approach is: due to we can not pass `nodeInfo` to extender thru HTTP, the real world extender can not do preemption logic since it doesn't even have knowledge of pods on the candiate node, how many resources been used/remaining etc.

Reply:
@resouer This assumption is not always correct. We track resources allocated for each pod in our extender implementation similar to how k8s tracks them. I spent some time today looking at the preemption implementation. I think the right place to invoke extender is right after the call to selectNodesForPreemption. The extender can potentially make this a smaller list if the pod doesn't fit in some of the nodes based on the resources it is tracking.

BTW, I am assuming you have a good use case for preemption with extenders. If you can describe your usage of extender, that would help also.

I will review the PR also.

Reply:
> I think the right place to invoke extender is right after the call to selectNodesForPreemption. The extender can potentially make this a smaller list if the pod doesn't fit in some of the nodes based on the resources it is tracking.

I wonder isn't this #56296 is doing?

Personally I do not have preemption use case in extender, but the current way of checking preemption is low efficient (one node at a time) and #56296 changed to sending nodes in batch to extender.

Reply:
@resouer 

#56296 is making the call at the right place, but it is using the existing Filter call which is incorrect per my earlier comment on that PR.

Reply:
OK, so it seems you prefer this PR's way of do preemption in extenders side, right? While I am not sure how user is expected to ensure the info in extender is consistent with default scheduler's cache? For example, when scheduler assigned a pod, the extender will not know it until that pod is delivered to etcd ... (That's why we abandoned this PR before since it seems extender can't do preemption)

Reply:
@resouer, given the assumptions made about extenders that they are expected to keep their own resource accounting, I agree with @ravigadde. We should send the nodes selected for preemption and their victims, plus the preemptor pod and let extenders shrink the list. Scheduler should do it only if the extenders support "preempt" verb. If they don't scheduler does not communicate with them (no more calling to "filter").

Reply:
> If they don't scheduler does not communicate with them (no more calling to "filter").

Emm, it's not clear how scheduler make sure this candidate node is able to run the preemptor pod? It could be rejected by a unpreemptable extender.

Reply:
>  It could be rejected by a unpreemptable extender.

That's a good point. In such cases, we may preempt pods without any benefit, but when scheduler tries to place the preemptor on the node, it will check with the extenders again and won't place the preemptor on a node that extenders don't approve. So, the placement of pods will still be correct, but scheduler does some useless preemptions. Unfortunately, I don't think we can do much about it when extenders do not support preemption and the feature is enabled.

Reply:
Thanks for detailed explaining. Since we have nominate annotation as precaution. Let's proceed with this approach. :smile_cat: